codegen/builtin_fs.go

```diff
@@ -281,9 +281,123 @@ func (f Frontend) Call(ctx context.Context, cln *client.Client, ret Register, op
 		}, solveOpts...)
 	})
 
+	err = g.Wait()
+	if err != nil {
+		return err
+	}
+
 	return ret.Set(fs)
 }
 
+func dockerfile(ctx context.Context, cln *client.Client, ret Register, opts Option, dt []byte) error {
+	var (
+		info        DockerfileInfo
+		solveOpts   []solver.SolveOption
+		sessionOpts []llbutil.SessionOption
+	)
+	for _, opt := range opts {
+		switch o := opt.(type) {
+		case DockerfileOption:
+			o(&info)
+		case solver.SolveOption:
+			solveOpts = append(solveOpts, o)
+		case llbutil.SessionOption:
+			sessionOpts = append(sessionOpts, o)
+		}
+	}
+
+	s, err := llbutil.NewSession(ctx, sessionOpts...)
+	if err != nil {
+		return err
+	}
+
+	g, ctx := errgroup.WithContext(ctx)
+
+	g.Go(func() error {
+		return s.Run(ctx, cln.Dialer())
+	})
+
+	zero := ZeroValue()
+	fs, err := zero.Filesystem()
+	if err != nil {
+		return err
+	}
+
+	g.Go(func() error {
+		var pw progress.Writer
+
+		mw := MultiWriter(ctx)
+		if mw != nil {
+			pw = mw.WithPrefix("", false)
+		}
+
+		return solver.Build(ctx, cln, s, pw, func(ctx context.Context, c gateway.Client) (res *gateway.Result, err error) {
+			caps := c.BuildOpts().LLBCaps
+			st, img, err := dockerfile2llb.Dockerfile2LLB(ctx, dt, dockerfile2llb.ConvertOpt{
+				Target:       info.Target,
+				MetaResolver: c,
+				// BuildArgs:         filter(opts, buildArgPrefix),
```

Comment:
I wish we could provide build args as arguments to the `fs` states exported by the Dockerfile. But because build args are unordered key/value pairs, I think we'd still need the same type of syntax that we use for frontend inputs, eg: 
```
dockerfile.someStage with option {
  arg "foo" "bar"
}
```

With a bit of hackery though, the Dockerfile stages could take `variadic string` args and they could be parsed as `key=value`, eg `dockerfile.someStage "foo=bar" "baz=bop"`

Reply:
Yeah I kept them commented because I thought we could provide a builtin `dockerBuild` that has options to expose these. This `func dockerfile` was meant to be consumed by both `DockerFileDecl` (for `hlb run` and `import`)  and `dockerBuild` builtin.